From 59f694022911538226c4abc877b6e9dbf059c162 Mon Sep 17 00:00:00 2001
From: Andy Whitcroft <apw@canonical.com>
Date: Thu, 26 Jul 2012 11:25:52 +0100
Subject: [PATCH 0035/1767] UBUNTU: SAUCE: (no-up) rds_ib_send() -- prevent
 local pings triggering BUG_ON()

Pining localhost on an infiniband connection can trigger a BUG_ON()
and cause a denial of service.

Fix identified by comparison of the RHEL source rpms.

CVE-2012-2372
BugLink: http://bugs.launchpad.net/bugs/1016299
Signed-off-by: Andy Whitcroft <apw@canonical.com>
Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 net/rds/ib_send.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/rds/ib_send.c b/net/rds/ib_send.c
index eac30bf..32bf501 100644
--- a/net/rds/ib_send.c
+++ b/net/rds/ib_send.c
@@ -502,7 +502,7 @@ int rds_ib_xmit(struct rds_connection *conn, struct rds_message *rm,
 	int flow_controlled = 0;
 	int nr_sig = 0;
 
-	BUG_ON(off % RDS_FRAG_SIZE);
+	BUG_ON(!conn->c_loopback && off % RDS_FRAG_SIZE);
 	BUG_ON(hdr_off != 0 && hdr_off != sizeof(struct rds_header));
 
 	/* Do not send cong updates to IB loopback */
-- 
1.9.3

